File: examples/java/BasicUsage.java
import com.skaidb.Skaidb;

/**
 * How to use skaidb from Java — modeled on JDBC.
 *
 * Compile & run (from this directory):
 *   javac -d out ../../drivers/java/src/main/java/com/skaidb/Skaidb.java BasicUsage.java
 *   java -cp out BasicUsage [host] [port] [user] [password]
 */
public class BasicUsage {
    public static void main(String[] args) {
        String host = args.length > 0 ? args[0] : "localhost";
        int port = args.length > 1 ? Integer.parseInt(args[1]) : 7000;
        String user = args.length > 2 ? args[2] : "anonymous";
        String pass = args.length > 3 ? args[3] : "";

        try (Skaidb.Connection conn = Skaidb.connect(host, port, user, pass)) {
            // --- DDL ---
            conn.execute("DROP TABLE IF EXISTS people");
            conn.execute("CREATE TABLE people (PRIMARY KEY (id))");

            // --- Prepared statement: parse once, bind (1-based, JDBC-style) per row ---
            try (Skaidb.Query ins = conn.prepare("INSERT INTO people (id, name, age) VALUES (?, ?, ?)")) {
                ins.setInt(1, 1).setString(2, "Ada").setInt(3, 36).executeUpdate();
                ins.setInt(1, 2).setString(2, "Linus").setInt(3, 54).executeUpdate();
                ins.setInt(1, 3).setString(2, "Margaret").setInt(3, 80).executeUpdate();
            }

            // --- Query ---
            System.out.println("age > 40:");
            try (Skaidb.Query q = conn.prepare("SELECT id, name, age FROM people WHERE age > ? ORDER BY id")) {
                Skaidb.ResultSet rs = q.setInt(1, 40).executeQuery();
                while (rs.next()) {
                    System.out.println("  " + rs.getObject("id") + " " + rs.getObject("name") + " " + rs.getObject("age"));
                }
            }

            // --- Update ---
            long updated;
            try (Skaidb.Query q = conn.prepare("UPDATE people SET age = ? WHERE id = ?")) {
                updated = q.setInt(1, 37).setInt(2, 1).executeUpdate();
            }
            System.out.println("updated " + updated + " row(s)");

            // --- Point read by primary key ---
            try (Skaidb.Query q = conn.prepare("SELECT name, age FROM people WHERE id = ?")) {
                Skaidb.ResultSet rs = q.setInt(1, 1).executeQuery();
                rs.next();
                System.out.println("id=1: " + rs.getObject("name") + " " + rs.getObject("age"));
            }

            // --- Error handling ---
            try {
                conn.execute("SELECT * FROM does_not_exist");
            } catch (Skaidb.SkaidbException e) {
                System.out.println("expected error: " + e.getMessage());
            }

            // --- Delete + cleanup ---
            long deleted;
            try (Skaidb.Query q = conn.prepare("DELETE FROM people WHERE id = ?")) {
                deleted = q.setInt(1, 2).executeUpdate();
            }
            System.out.println("deleted " + deleted + " row(s)");
            conn.execute("DROP TABLE people");
        }
    }
}

File: examples/php/basic_usage.php
<?php

declare(strict_types=1);

/**
 * How to use skaidb from PHP — modeled on PDO.
 *
 *     php basic_usage.php [host] [port] [user] [password]
 *
 * Uses the driver at ../../drivers/php/src. In a real project, install it
 * via Composer instead of the require below, which exists only so this
 * example runs straight out of the repo.
 */

require __DIR__ . '/../../drivers/php/src/Skaidb.php';

use Skaidb\Connection;
use Skaidb\SkaidbException;

$host = $argv[1] ?? 'localhost';
$port = isset($argv[2]) ? (int) $argv[2] : 7000;
$user = $argv[3] ?? 'anonymous';
$password = $argv[4] ?? '';

try {
    $db = new Connection($host, $port, $user, $password, 'QUORUM');

    // --- DDL ---
    try {
        $db->exec('DROP TABLE people');
    } catch (SkaidbException $e) {
        // table didn't exist — fine
    }
    $db->exec('CREATE TABLE people (PRIMARY KEY (id))');

    // --- Batch insert with a prepared statement (`?` placeholders) ---
    $insert = $db->prepare('INSERT INTO people (id, name, age) VALUES (?, ?, ?)');
    foreach ([[1, 'Ada', 36], [2, 'Linus', 54], [3, 'Margaret', 80]] as $row) {
        $insert->execute($row);
    }

    // --- Query ---
    $stmt = $db->prepare('SELECT id, name, age FROM people WHERE age > ? ORDER BY id');
    $stmt->execute([40]);
    echo "age > 40:\n";
    foreach ($stmt->fetchAll() as $person) {
        printf("  %d %s %d\n", $person['id'], $person['name'], $person['age']);
    }

    // --- Update ---
    $upd = $db->prepare('UPDATE people SET age = ? WHERE id = ?');
    $upd->execute([37, 1]);
    echo 'updated ' . $upd->rowCount() . " row(s)\n";

    // --- Point read by primary key ---
    $one = $db->prepare('SELECT name, age FROM people WHERE id = ?');
    $one->execute([1]);
    $row = $one->fetch();
    echo "id=1: {$row['name']} {$row['age']}\n";

    // --- Error handling ---
    try {
        $db->exec('SELECT * FROM does_not_exist');
    } catch (SkaidbException $e) {
        echo 'expected error: ' . $e->getMessage() . "\n";
    }

    // --- Delete + cleanup ---
    $del = $db->prepare('DELETE FROM people WHERE id = ?');
    $del->execute([2]);
    echo 'deleted ' . $del->rowCount() . " row(s)\n";
    $db->exec('DROP TABLE people');

    $db->close();
} catch (SkaidbException $e) {
    fwrite(STDERR, 'skaidb error: ' . $e->getMessage() . "\n");
    exit(1);
}
